test/source-updater.test.js

```diff
@@ -1,232 +1,769 @@
+        done();
+      });
+    });
+  });
+});
+
+QUnit.test('audioQueueCallback calls callback immediately if queue is empty',
+function(assert) {
+  // Source buffer must exist for the callback to run. This case isn't tested, as it isn't
+  // required behavior (at the moment), but is necessary to know for this test.
+  this.sourceUpdater.createSourceBuffers({
+    audio: 'mp4a.40.2'
+  });
+
+  let executedCallback = false;
 
-  this.mediaSource.trigger('sourceopen');
+  this.sourceUpdater.audioQueueCallback(() => {
+    executedCallback = true;
+  });
 
-  assert.equal(this.mediaSource.sourceBuffers.length, 1, 'created one source buffer');
-  assert.equal(this.mediaSource.sourceBuffers[0].mimeType_, 'video/mp2t',
-              'assigned the correct MIME type');
+  assert.ok(executedCallback, 'executed callback');
 });
 
-QUnit.test('runs a callback when the source buffer is created', function(assert) {
-  let updater = new SourceUpdater(this.mediaSource, 'video/mp2t');
-  let sourceBuffer;
+QUnit.test('videoQueueCallback calls callback immediately if queue is empty',
+function(assert) {
+  // Source buffer must exist for the callback to run. This case isn't tested, as it isn't
+  // required behavior (at the moment), but is necessary to know for this test.
+  this.sourceUpdater.createSourceBuffers({
+    video: 'avc1.4D001E'
+  });
+
+  let executedCallback = false;
 
-  updater.appendBuffer(new Uint8Array([0, 1, 2]));
+  this.sourceUpdater.videoQueueCallback(() => {
+    executedCallback = true;
+  });
 
-  this.mediaSource.trigger('sourceopen');
-  sourceBuffer = this.mediaSource.sourceBuffers[0];
-  assert.equal(sourceBuffer.updates_.length, 1, 'called the source buffer once');
-  assert.deepEqual(sourceBuffer.updates_[0].append, new Uint8Array([0, 1, 2]),
-                  'appended the bytes');
+  assert.ok(executedCallback, 'executed callback');
 });
 
-QUnit.test('runs callback if a media source exists when passed source buffer emitter',
+QUnit.test('audioQueueCallback calls callback after queue empties if queue is not empty',
 function(assert) {
-  let sourceBufferEmitter = new videojs.EventTarget();
-  let sourceBuffer;
+  const done = assert.async();
+
+  // Source buffer must exist for the callback to run. This case isn't tested, as it isn't
+  // required behavior (at the moment), but is necessary to know for this test.
+  this.sourceUpdater.createSourceBuffers({
+    audio: 'mp4a.40.2'
+  });
 
-  this.mediaSource.trigger('sourceopen');
-  // create other media source
-  this.mediaSource.addSourceBuffer('audio/mp2t');
+  let executedCallback = false;
+  let appendedAudio = false;
 
-  let updater = new SourceUpdater(this.mediaSource, 'video/mp2t', sourceBufferEmitter);
+  this.sourceUpdater.appendBuffer('audio', mp4Audio(), () => {
+    appendedAudio = true;
+    assert.notOk(executedCallback, 'haven\'t executed callback');
+    setTimeout(() => {
```

Comment:
is there another way to ensure that the queue cleared instead of a timeout?

Reply:
We're only using a timeout here to ensure we can run something right after the callback finishes and is marked as called. Unless we add another event, I think we have to use the timeout.

Reply:
Could we also use the sinon clock?